zk/stages/stage_witness.go

```diff
@@ -114,149 +104,91 @@ func SpawnStageWitness(
+	if fromBatch < highestVerifiedBatchWithOffset {
+		fromBatch = highestVerifiedBatchWithOffset
 	}
 
-	reader := state.NewPlainState(tx, blocks[0].NumberU64(), systemcontracts.SystemContractCodeLookup[cfg.chainConfig.ChainName])
-	defer reader.Close()
-	prevStateRoot := prevHeader.Root
-
-	log.Info(fmt.Sprintf("[%s] Executing blocks and collecting witnesses", logPrefix), "from", startBlock, "to", stageInterhashesProgressBlockNo)
-
-	now := time.Now()
-
-	// used to ensure that any info tree updates for this batch are included in the witness - re-use of an index for example
-	// won't write to storage so will be missing from the witness but the prover needs it
-	forcedInfoTreeUpdates := make([]common.Hash, 0)
-
-	for _, block := range blocks {
-		reader.SetBlockNr(block.NumberU64())
-		tds := state.NewTrieDbState(prevHeader.Root, tx, startBlock-1, nil)
-		tds.SetResolveReads(true)
-		tds.StartNewBuffer()
-		tds.SetStateReader(reader)
-
-		trieStateWriter := tds.NewTrieStateWriter()
-		if err := witness.PrepareGersForWitness(block, memHermezDb, tds, trieStateWriter); err != nil {
-			return fmt.Errorf("PrepareGersForWitness: %w", err)
-		}
-
-		getHashFn := core.GetHashFn(block.Header(), getHeader)
-
-		chainReader := stagedsync.NewChainReaderImpl(cfg.chainConfig, tx, nil, log.New())
-
-		vmConfig := vm.Config{}
-		if _, err = core.ExecuteBlockEphemerallyZk(cfg.chainConfig, &vmConfig, getHashFn, cfg.engine, block, tds, trieStateWriter, chainReader, nil, hermezDb, &prevStateRoot); err != nil {
-			return fmt.Errorf("ExecuteBlockEphemerallyZk: %w", err)
-		}
-
-		forcedInfoTreeUpdate, err := witness.CheckForForcedInfoTreeUpdate(memHermezDb, block.NumberU64())
-		if err != nil {
-			return fmt.Errorf("CheckForForcedInfoTreeUpdate: %w", err)
-		}
-		if forcedInfoTreeUpdate != nil {
-			forcedInfoTreeUpdates = append(forcedInfoTreeUpdates, *forcedInfoTreeUpdate)
-		}
-
-		prevStateRoot = block.Root()
-
-		w, err := witness.BuildWitnessFromTrieDbState(ctx, memTx, tds, reader, cfg.forcedContracs, forcedInfoTreeUpdates, false)
-		if err != nil {
-			return fmt.Errorf("BuildWitnessFromTrieDbState: %w", err)
-		}
+	hermezDb := hermez_db.NewHermezDb(tx)
+	g := witness.NewGenerator(cfg.dirs, cfg.historyV3, cfg.agg, cfg.blockReader, cfg.chainConfig, cfg.zkCfg, cfg.engine, cfg.forcedContracts, cfg.unwindLimit)
 
-		bytes, err := witness.GetWitnessBytes(w, false)
+	for batchNo := fromBatch; batchNo <= highestVerifiedBatch.BatchNo; batchNo++ {
+		badBatch, err := reader.GetInvalidBatch(batchNo)
 		if err != nil {
-			return fmt.Errorf("GetWitnessBytes: %w", err)
+			return fmt.Errorf("GetInvalidBatch: %w", err)
 		}
 
-		if hermezDb.WriteWitnessCache(block.NumberU64(), bytes); err != nil {
-			return fmt.Errorf("WriteWitnessCache: %w", err)
-		}
-		if time.Since(now) > 10*time.Second {
-			log.Info(fmt.Sprintf("[%s] Executing blocks and collecting witnesses", logPrefix), "block", block.NumberU64())
-			now = time.Now()
+		if !badBatch {
```

Comment:
maybe invert this `if` to keep nesting a little easier to follow and use a `continue` for the loop